torch/csrc/dynamo/python_compiled_autograd.cpp

```diff
@@ -0,0 +1,479 @@
+        result.emplace_back(
+            py::cast<c10::SymInt>(PyList_GET_ITEM(pyresult, idx++)));
+      } else {
+        result.emplace_back();
+      }
+    }
+    TORCH_CHECK(idx == result_len && result.size() == expected_sizes.size());
+    return result;
+  }
+
+  // TODO(jansel): benchmark map vs unordered_map
+  std::unordered_map<CacheKey, std::unique_ptr<CacheNode>> next;
+  std::vector<CacheKeyBuffer> key_storage;
+  std::vector<SizeInput> expected_sizes;
+  THPObjectPtr compiled_fn;
+};
+
+struct InputBuffers : public std::unordered_map<Node*, InputBuffer> {
+  InputBuffer& lookup(Node* function) {
+    auto it = find(function);
+    if (it == end()) {
+      it = emplace(function, InputBuffer(function->num_inputs())).first;
+    }
+    return it->second;
+  }
+};
+
+static PyObject* the_autograd_compiler = nullptr;
+static PyObject* set_autograd_compiler(PyObject* dummy, PyObject* args);
+
+static PyObject* clear_cache(PyObject* dummy, PyObject* args) {
+  CacheNode::root()->clear();
+  Py_RETURN_NONE;
+}
+
+static PyObject* is_cache_empty(PyObject* dummy, PyObject* args) {
+  if (CacheNode::root()->is_empty()) {
+    Py_RETURN_TRUE;
+  }
+  Py_RETURN_FALSE;
+}
+
+static PyMethodDef _methods[] = {
+    {"set_autograd_compiler", set_autograd_compiler, METH_VARARGS, NULL},
+    {"clear_cache", clear_cache, METH_NOARGS, NULL},
+    {"is_cache_empty", is_cache_empty, METH_NOARGS, NULL},
+    {NULL, NULL, 0, NULL}};
+
+static struct PyModuleDef _module = {
+    PyModuleDef_HEAD_INIT,
+    "torch._C._dynamo.autograd_compiler",
+    "Hooks for compiling autograd",
+    -1,
+    _methods};
+
+static TraceState call_begin_capture(
+    PyObject* self,
+    CacheNode& cache,
+    AutogradCompilerCall& compiler_call,
+    size_t num_outputs) {
+  static PyObject* method_name = PyUnicode_InternFromString("begin_capture");
+  THPObjectPtr pyinput(THPVariable_WrapList(compiler_call.inputs));
+  THPObjectPtr pysizeinput(cache.wrap_dynamic_inputs());
+  THPObjectPtr pyresult(check(PyObject_CallMethodObjArgs(
+      self, method_name, pyinput.get(), pysizeinput.get(), NULL)));
+
+  PyObject *fake_inputs, *fake_sizes;
+  check(PyArg_ParseTuple(pyresult.get(), "OO", &fake_inputs, &fake_sizes));
+  return TraceState(
+      THPVariable_UnpackList(fake_inputs),
+      cache.unwrap_dynamic_inputs(fake_sizes),
+      compiler_call.accumulate_grad,
+      num_outputs);
+}
+
+static PyObject* call_end_capture(PyObject* self, const variable_list& inputs) {
+  static PyObject* method_name = PyUnicode_InternFromString("end_capture");
+  THPObjectPtr pyinput(THPVariable_WrapList(inputs));
+  return check(PyObject_CallMethodOneArg(self, method_name, pyinput.get()));
+}
+
+variable_list compiled_autograd(
+    const std::shared_ptr<Node>& graph_root,
+    GraphTask& graph_task,
+    bool accumulate_grad,
+    const edge_list& output_edges) {
+  TORCH_CHECK(
+      output_edges.empty() || !accumulate_grad,
+      "outputs+accumulate_grad not yet implemented")
```

Comment:
```suggestion
      "specifying inputs= with .backward() not yet implemented")
```

Reply:
edit: "specifying inputs= with .backward() not yet implemented **for compiled autograd**"